// @dojo/cli-build-app/base.test.config.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const base_config_1 = require("./base.config");
const WrapperPlugin = require('wrapper-webpack-plugin');
function webpackConfig(args) {
    const config = base_config_1.default(args);
    const { plugins, module } = config;
    const externals = config.externals || [];
    const instrumenterOptions = args.legacy ? {} : { esModules: true };
    config.plugins = [
        ...plugins,
        new WrapperPlugin({
            test: /(all.*(\.js$))/,
            footer: `typeof define === 'function' && define.amd && define(['${base_config_1.libraryName}'], function() {});`
        })
    ];
    if (module) {
        module.rules = module.rules.map((rule) => {
            if (Array.isArray(rule.use)) {
                rule.use = rule.use.map((loader) => {
                    if (typeof loader === 'string') {
                        return loader;
                    }
                    const { loader: loaderName } = loader;
                    if (loaderName === 'umd-compat-loader') {
                        return {
                            loader: loaderName,
                            options: {}
                        };
                    }
                    return loader;
                });
            }
            return rule;
        });
        module.rules.push({
            test: /src[\\\/].*\.ts(x)?$/,
            use: {
                loader: 'istanbul-instrumenter-loader',
                options: instrumenterOptions
            },
            enforce: 'post'
        });
    }
    externals.push(/^intern/);
    config.externals = externals;
    config.devtool = 'inline-source-map';
    return config;
}
exports.default = webpackConfig;
